src/core/configs/cluster.rs

```diff
@@ -0,0 +1,397 @@
+            != 0
+        {
+            debug!("config contains duplicates vdisks ids");
+            return Some("config contains duplicates vdisks ids".to_string());
+        }
+        if self
+            .nodes
+            .as_ref()?
+            .iter()
+            .group_by(|x| x.name.clone())
+            .into_iter()
+            .map(|(_, group)| group.count())
+            .filter(|x| *x > 1)
+            .count()
+            != 0
+        {
+            debug!("config contains duplicates nodes names");
+            return Some("config contains duplicates nodes names".to_string());
+        }
+
+        for node in self.nodes.as_ref()?.iter() {
+            node.prepare();
+        }
+
+        for vdisk in self.vdisks.as_ref()?.iter() {
+            for replica in vdisk.replicas.as_ref()?.iter() {
+                match self.nodes.as_ref()?.iter().find(|x| x.name == replica.node) {
+                    Some(node) => {
+                        if node.disks.as_ref()?.iter().find(|x| x.name == replica.disk) == None {
+                            debug!(
+                                "cannot find in node: {}, disk with name: {} for vdisk: {}",
+                                replica.node.as_ref()?,
+                                replica.disk.as_ref()?,
+                                vdisk.id?
+                            );
+                            return Some(format!(
+                                "cannot find in node: {}, disk with name: {} for vdisk: {}",
+                                replica.node.as_ref()?,
+                                replica.disk.as_ref()?,
+                                vdisk.id?
+                            ));
+                        }
+                    }
+                    None => {
+                        debug!(
+                            "cannot find node: {} for vdisk: {}",
+                            replica.node.as_ref()?,
+                            vdisk.id?
+                        );
+                        return Some(format!(
+                            "cannot find node: {} for vdisk: {}",
+                            replica.node.as_ref()?,
+                            vdisk.id?
+                        ));
+                    }
+                }
+            }
+        }
+
+        None
+    }
+}
+
+use crate::core::data::Node as DataNode;
+use crate::core::data::NodeDisk as DataNodeDisk;
+use crate::core::data::VDisk as DataVDisk;
+
+use std::collections::HashMap;
+
+pub trait BobConfig {
+    fn get_cluster_config(&self, filename: &String) -> Result<Vec<DataVDisk>, String>;
+
+    fn read_config(&self, filename: &String) -> Result<Cluster, String>;
+    fn parse_config(&self, config: &String) -> Result<Cluster, String>;
+    fn convert_to_data(&self, cluster: &Cluster) -> Option<Vec<DataVDisk>>;
+}
+
+pub struct YamlConfig {}
+
+impl BobConfig for YamlConfig {
+    fn read_config(&self, filename: &String) -> Result<Cluster, String> {
```

Comment:
More convenient to use `&str`, because &String means, that you have to pass reference to object String. But if you want to use `static str`:
```rust
let s = "test static str";
```
 as filename, you will have to create new owned `String` first: 
```rust
config.read_config(&s.to_string());
```
. But in case of `&str` it would be only: 
```rust
config.read_config(s);
```